### DISCUSSION TITLE
HX711 module with Raspberry Pi Pico, raw values stay the same even if I add weight

### DISCUSSION THREAD

**Discussion by FrankHenckens** [micropython/micropython#15896] [Core Development]
Hello,

I am making a scale with a bending beam load cell, a HX711 and a Raspberry Pi Pico. I read the raw values of the loadcell and when I add weight, the raw value stays the same. It fluctuates but it also fluctuates without any added weight. All help is welcome.

![WhatsApp Image 2024-09-23 at 10 28 27](https://github.com/user-attachments/assets/d5d33dc0-f923-4796-8a75-0651151facfe)
![WhatsApp Image 2024-09-23 at 10 28 28](https://github.com/user-attachments/assets/a13e24de-f4db-416e-8a77-88fd90ff1b56)
![WhatsApp Image 2024-09-23 at 10 28 29](https://github.com/user-attachments/assets/d2224faf-487c-4911-8cd8-c95cfd74c024)

**Comment by robert-hh** [comment]
If there is no change, either

1. the load cell is broken
2. the connection to the load cell is wrong or
3. you use the wrong input of the HX711.

The color scheme of the document is unusual. Usually, red and black ate V+ and V-, and green and white are the output wires.  If you have a multimeter, you can check the impedance between the wires of the load cell. The impedance between V+ and V+ should be similar to the impedance between the output wires, and the impedance between V+ (or V-) and each output wire about half of the one between V+ and V-.
With a gain setting of 128 and 64 you have to use input A, for a gain of 32 input B.
The HX711 is quite noisy. Usually at least the lower 10 bits of the result vary.

**Reply by FrankHenckens** [reply]
The load cell is not broken. I know the color scheme is unusual, I made the mistake of connecting it wrong before. I am using a gain of 128, so I use A. When I measure the voltage in mV on A+ and A-, the voltage is correct. Its connected to the 5V of the pico, the E+ and E- measure 4,15V. Its a 250Kg load cell with 2mV/V. So with a load of 250Kg I should get a milivoltage of 4,15*2=8,3mV. When I add arround 20Kg I measure 0,7mV on A. 8,3 / 250 * 20 = 0,7mV.

**Reply by robert-hh** [reply]
a) These voltages are somewhat strange. Usually load cells are symmetric. So at 5V the output levels at A+ and A- to GND should be about V+ / 2. 
b) According to the HX711 data sheet he common mode range for A+/A- is AGND+1.2V to AVCC-1.3V. So at V+ of 5V the permitted range for A+ and A-  is 1.2V to 3.7V. V+ at the load cell may be even lower than 5V due to an internal regulator. A voltage of 4.15 V to GND at A+/A- is definitely out of the common mode range.

**Reply by FrankHenckens** [reply]
the 4,15V is between E+ and E-, the voltage  between a+ and A- is 0,7mV when I add 20Kg to the load cell.

**Reply by robert-hh** [reply]
what is the voltage between A+ and GND?

**Reply by FrankHenckens** [reply]
1,98V

**Reply by robert-hh** [reply]
That's fine. 
I see that you run the HX711 board at 5V. That means that the CLK and DOUT logig levels are at 5V too. That does not match the RPi Pico digital levels, which are at 3.3V. It may work, but is definitely not recommended and may break the RP2. At least you should add a resistor like 4k7  Ohm between the DOUT of the HX711 and the RP2 GPIO pin to limit the current that has to absorbed by the GPIO input surge protection. Better run the HX711 at a lower voltage. But I do not know if it works at 3.3V. May be too low due to the internal regulator.
Which software driver do you use for the HX711?

**Reply by FrankHenckens** [reply]
I have al logic level converter to switch the 5V from the HX711 to 3,3V for the pico, but I saw online it was not necessary. 

The library I have: 
from machine import Pin
import utime

class HX711:
    def __init__(self, dout, pd_sck, gain=128):
        self.dout = Pin(dout, Pin.IN, pull=Pin.PULL_DOWN)
        self.pd_sck = Pin(pd_sck, Pin.OUT)
        self.pd_sck.off()

        if gain == 128:
            self.GAIN = 1
        elif gain == 64:
            self.GAIN = 3
        elif gain == 32:
            self.GAIN = 2

        self.read()
        self.offset = 0  # Initialize offset to 0

    def is_ready(self):
        return self.dout.value() == 0

    def read(self):
        while not self.is_ready():
            pass

        result = 0
        for i in range(24):
            self.pd_sck.on()
            result = result << 1
            self.pd_sck.off()
            if self.dout.value():
                result += 1

        self.pd_sck.on()
        result = result ^ 0x800000
        self.pd_sck.off()

        for i in range(self.GAIN):
            self.pd_sck.on()
            self.pd_sck.off()

        return result

    # Voeg de tare functie toe om de weegschaal naar 0 te resetten
    def tare(self, times=10):
        sum = 0
        for _ in range(times):
            sum += self.read()  # Neem meerdere metingen
            utime.sleep_ms(10)
        self.offset = sum / times  # Bereken en stel het gemiddelde in als offset

    # Functie om het gewicht te krijgen
    def get_units(self, times=1):
        total = 0
        for _ in range(times):
            total += self.read()
            utime.sleep_ms(10)
        return (total / times) - self.offset  # Gebruik de offset om het resultaat aan te passen

    def power_down(self):
        self.pd_sck.off()
        self.pd_sck.on()
        utime.sleep_ms(1)

    def power_up(self):
        self.pd_sck.off()
        utime.sleep_ms(50)

My main code:
#libraries importeren
from machine import I2C, Pin
import utime
from hx711 import HX711
from lcd_api import LcdApi
from pico_i2c_lcd import I2cLcd

#adres bekend maken & display afmetingen
I2C_ADDR     = 0x27
I2C_NUM_ROWS = 2
I2C_NUM_COLS = 16

#pinnen van i2c backpack & welke bus + frequentie
i2c = I2C(1, sda=machine.Pin(2), scl=machine.Pin(3), freq=400000)
lcd = I2cLcd(i2c, I2C_ADDR, I2C_NUM_ROWS, I2C_NUM_COLS)

#pinnen hx711 module
hx = HX711(dout=Pin(4), pd_sck=Pin(5), gain=64)
hx.tare()
print("Tare uitgevoerd. Offset ingesteld op: ", hx.offset)

#OpstartDisplay
def greeting():
    lcd.clear()
    lcd.move_to(6,0)
    lcd.putstr("H&F")
    lcd.move_to(2,1)
    lcd.putstr("Electronics")
    utime.sleep(2)
    lcd.clear()
    utime.sleep(1)
    lcd.move_to(5,0)
    lcd.putstr("Zeugen")
    lcd.move_to(2,1)
    lcd.putstr("Weeg Systeem")
    utime.sleep(3)
    lcd.clear()
    utime.sleep(1)
    
# Functie om gewicht te lezen en weer te geven
def display_weight():
    try:
        weight = hx.get_units()  # Lees de waarde van de HX711 uit
        scaled_weight = (weight / 1)  #gewicht delen door schaalfactor
        #if scaled_weight < 0:
         #   scaled_weight = 0 #wanneer het kleiner is dan 0, maak het 0, zodat er geen - komt
        lcd.clear()
        lcd.move_to(0, 0)
        lcd.putstr("Gewicht:")
        lcd.move_to(9, 0)
        lcd.putstr("{0:.1f}Kg".format(scaled_weight)) #geef het gewicht weer met 1 decimalen en Kg erachter op display
        print("Gewicht: {0:.2f} Kg".format(scaled_weight))#geef het gewicht weer met 2 decimalen en Kg erachter in shell
        lcd.move_to(0, 1)
        lcd.putstr("{0:.0f}".format(weight)) #geef het gewicht weer met 0 decimalen en Kg erachter op display
        print("Gewicht: {0:.2f} ".format(weight))#geef het gewicht weer met 2 decimalen en Kg erachter in shell
    except OSError as e:
        lcd.clear()
        lcd.putstr("Error: HX711")
        print("Error: ", e)
        
# Mainloop
while True:
    display_weight()
    utime.sleep(2)

**Reply by FrankHenckens** [reply]
I tried get_units and also read

**Reply by robert-hh** [reply]
The HX711 driver looks similar to the one I use and which works. Some differences:

- the unconditional `result = result ^ 0x800000` in the `hx.read()` method looks wrong. It inverts always bit 23.
- the method `is_ready()` is not fully robust if written this way. The HX711 has the habit of putting a conversion-done pulse at the DOUT line every time a conversion is done, even it data is read at that moment. In that case, the data is corrupted.

Besides that, which values do you get as result from hx.read()?

You find the driver I use at https://github.com/robert-hh/hx711. With the RP2 you can use the GPIO or PIO variant. The PIO variant has a more precises timing. Both take care of the conversion done pulse.

**Reply by FrankHenckens** [reply]
from hx.read I get the value 8384017 and it fluctuates

**Reply by robert-hh** [reply]
Which is 0x7fee11. Taking into acount the unconditional xor, the value read from the sensor was 0xffee1, which should be proprly converted to -4591.

**Reply by robert-hh** [reply]
Another comment about your hx.read() method. It makes one clock pulse more than needed. For GAIN=128, it makes 26 clock pulses instead of 25, for GAIN=64 it makes 28 clock pulses instead of 27. 
if you do not want to try the code from https://github.com/robert-hh/hx711, change at least:

```
    self.pd_sck.on()
    result = result ^ 0x800000
    self.pd_sck.off()
```
to 
```
    if result > 0x7fffff:
        result -= 0x1000000
```

Note: When posting code, emclose the code block into line consisting of three backticks ```.

**Reply by FrankHenckens** [reply]
I have your library and it works perfect, thank you!

**Reply by robert-hh** [reply]
Very good. Thank you for your feedback.

**Reply by FrankHenckens** [reply]
I have one more question, are the logic leven converters necessary in my circuit? I thought it was necessary between the HX711 modules and the pico, but on the internet I see people doing it without logic level converters.

**Reply by robert-hh** [reply]
Software usually does not change when not touched. Hardware can, if you move it. 0 as result is strange. It means that data or clock line do not work. Check the wires. It may be related to the 5V/3.3V difference.

Besides that I would recommend with an RP2 to use the pio version. The clock timing is more precise.

About the logic level converters, there are two aspects:

a) RP2 outgoing signals (CLK). The output level of the RP2 are 0 and about 3.2 V. The high level of the HX711 CLK input at 5V is not specified. For a CMOS circuitry the guaranteed high level is usually 3 - 3.5V. So it's in a gray zone, whether the high level of a clock is seen as such by the HX711 and may vary with the temperature.

b) RP2 input.  The maximum input voltage at the GPIO pin is 3.6V for a 3.3V supply. The HX711 outputs about 5V with a limited current, which is not specified. That excess current has to be consumed by the surge protection diodes of the GPIO pad of the RP2, but they not designed for that. You could limit the current by inserting a series resistor in the data line. 4700 Ohm would limit the current to <300µA. Care must be taken to reduce the bandwidth of the connection.

But better use a level converter. Since it does not have to be open drain, something based e.g. on a TXB010x (like txb0102 or txb0104) would be the best. The offer better speed.

**Reply by robert-hh** [reply]
About levels: You could try to reduce the supply voltage of the HX711. It runs at 3.3V, only the input voltage range is reduced. At a GAIN of 128 it would be about 20mV. You may have to bypass the voltage regulator on the HX711 board.

**Comment by bidrohini** [comment]
I suggest cross-checking the loadcells in any other platform. If you have a full-fledged Raspberry Pi, you can try to test the load cells with that. Here is a detailed tutorial and sample code: https://www.theengineeringprojects.com/2023/06/interface-weight-sensor-hx711-with-raspberry-pi-4.html
If you have an Arduino, you can test the load cells with that too with relevant codes. If the cells work in other platforms, That will mean that problem the is in your pico setup or code.

**Reply by robert-hh** [reply]
@bidrohini The code in the example you've linked is wrong:

- It does not handle the sign bit properly. It inverts unconditionally the sign bit. From the data sheet: _The output 24 bits of data is in 2’s complement format._ The output range is 0x800000 to 0x0x7fffff. So if you invert the sign,  for instance a value of 0x800000 (-8388608) gets 0. And that's truly wrong.
- It does not handle the `conversion ready` flag, causing spurious out-of-range results.
- It does not care for the clk pulse to be shorter than 50µs. If its longer, the device will fall into sleep until the next clock pulse and then wake up again for the remainder of the read transaction, which then is corrupted.

**Reply by FrankHenckens** [reply]
Hello Robert,

I would love your help with another project of mine if you have time. https://github.com/orgs/micropython/discussions/15947
